main.c

```diff
@@ -111,6 +118,80 @@ struct tstatstruct {
 VEC_STRUCT(tstatsvec,struct tstatstruct);
 #endif
 
+typedef union {
+	u8 raw[pkprefixlen + PUBLIC_LEN + 32];
+	struct {
+		u64 prefix[4];
+		u64 key[4];
+		u64 hash[4];
+	} i;
+} pubonionunion;
+
+
+
+#define BUF_APPEND(buf,offset,src,srclen) strncpy(&buf[offset],src,srclen); offset += srclen;
+#define BUF_APPEND_CSTR(buf,offset,src) BUF_APPEND(buf,offset,src,strlen(src))
+#define BUF_APPEND_CHAR(buf,offet,c) buf[offset++] = c;
+
+static void writekeys(const char *hostname, const u8 *formated_secret, const u8 *formated_public)
+{
+	char keysbuf[KEYS_LEN];
+	size_t offset = 0;
+
+	BUF_APPEND_CSTR(keysbuf, offset, keys_field_generated);
+	BUF_APPEND_CHAR(keysbuf, offset, '\n');
+
+	BUF_APPEND_CSTR(keysbuf, offset, keys_field_hostname);
+	BUF_APPEND(keysbuf, offset, hostname, ONION_LEN);
+	BUF_APPEND_CHAR(keysbuf, offset, '\n');
+
+	BUF_APPEND_CSTR(keysbuf, offset, keys_field_secretkey);
+	char seckeybuf[SECKEY_LEN + NULLTERM_LEN];
+	base64_to(seckeybuf, formated_secret, FORMATTED_SECRET_LEN);
+	BUF_APPEND(keysbuf, offset, seckeybuf, SECKEY_LEN);
+	BUF_APPEND_CHAR(keysbuf, offset, '\n');
+
+	BUF_APPEND_CSTR(keysbuf, offset, keys_field_publickey);
+	char pubkeybuf[PUBKEY_LEN + NULLTERM_LEN];
+	base64_to(pubkeybuf, formated_public, FORMATTED_PUBLIC_LEN);
+	BUF_APPEND(keysbuf, offset, pubkeybuf, PUBKEY_LEN);
+	BUF_APPEND_CHAR(keysbuf, offset, '\n');
+
+	BUF_APPEND_CSTR(keysbuf, offset, keys_field_time);
+	char timebuf[TIME_LEN + NULLTERM_LEN];
+	time_t timer;
+	struct tm* tm_info;
+	time(&timer);
+	tm_info = localtime(&timer);
```

Comment:
http://yaml.org/type/timestamp.html
>If the time zone is omitted, the timestamp is assumed to be specified in UTC. The time part may be omitted altogether, resulting in a date format. In such a case, the time part is assumed to be 00:00:00Z (start of day, UTC).

Reply:
actually, `localtime` and `gmtime` both use internal shared object and access to them must be synchronised (they aren't thread safe).
https://stackoverflow.com/questions/20820384/a-thread-safe-version-of-the-standard-c-function-localtime

Reply:
Oh, my... Yes, thread safety is violated and timezone is not mentioned. Second is simple (I hope).
About thread safety... If it's possible to convert `time_t` to textual representation in UTC in a portable and thread safe manner -- this is the way to go (why is this not in standard yet?!).

Right now:
1. We can wrap this place by another mutex. Usually we find something not so often, so overhead of locking mutex once a few seconds is invisible.
2. If somebody has monstrosity hardware and catching tens thousand of addresses per second, then we can take current time in main thread, convert it into c-string, put *atomically* to some pointer available to working threads, and get it *atomically* there with *atomic* count referencing. Kind of complicated but bullet-proof for performance and can be done correctly.

Reply:
I guess for now we should just lock it with additional mutex.
Don't think anyone outputs onions that fast, I kinda focused most of my optimization effort in generation/filtering, because most of time is spent there.
we already lock file handle we output too.

Reply:
>why is this not in standard yet?!
dunno. there's `localtime_r` and `gmtime_r` but they are GNU and BSD specific. :\